rust2/core/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+pub trait Statement: Optionable<Option = OptionStatement> {
+    /// Bind Arrow data. This can be used for bulk inserts or prepared
+    /// statements.
+    fn bind(&self, batch: RecordBatch) -> Result<()>;
+
+    /// Bind Arrow data. This can be used for bulk inserts or prepared
+    /// statements.
+    fn bind_stream(&self, reader: Box<dyn RecordBatchReader + Send>) -> Result<()>;
+
+    /// Execute a statement and get the results.
+    ///
+    /// This invalidates any prior result sets.
+    fn execute(&self) -> Result<impl RecordBatchReader + Send>;
+
+    /// Execute a statement that doesn’t have a result set and get the number
+    /// of affected rows.
+    ///
+    /// This invalidates any prior result sets.
+    ///
+    /// # Result
+    ///
+    /// Will return the number of rows affected. If the affected row count is
+    /// unknown or unsupported by the database, will return `Ok(-1)`.
+    fn execute_update(&self) -> Result<i64>;
+
+    /// Get the schema of the result set of a query without executing it.
+    ///
+    /// This invalidates any prior result sets.
+    ///
+    /// Depending on the driver, this may require first executing
+    /// [Statement::prepare].
+    ///
+    /// # Since
+    ///
+    /// ADBC API revision 1.1.0
+    fn execute_schema(&self) -> Result<Schema>;
+
+    /// Execute a statement and get the results as a partitioned result set.
+    fn execute_partitions(&self) -> Result<PartitionedResult>;
+
+    /// Get the schema for bound parameters.
+    ///
+    /// This retrieves an Arrow schema describing the number, names, and
+    /// types of the parameters in a parameterized statement. The fields
+    /// of the schema should be in order of the ordinal position of the
+    /// parameters; named parameters should appear only once.
+    ///
+    /// If the parameter does not have a name, or the name cannot be
+    /// determined, the name of the corresponding field in the schema will
+    /// be an empty string. If the type cannot be determined, the type of
+    /// the corresponding field will be NA (NullType).
+    ///
+    /// This should be called after [Statement::prepare].
+    fn get_parameters_schema(&self) -> Result<Schema>;
+
+    /// Turn this statement into a prepared statement to be executed multiple
+    /// times.
+    ///
+    /// This invalidates any prior result sets.
+    fn prepare(&self) -> Result<()>;
+
+    /// Set the SQL query to execute.
+    ///
+    /// The query can then be executed with [Statement::execute]. For queries
+    /// expected to be executed repeatedly, call [Statement::prepare] first.
+    fn set_sql_query(&self, query: &str) -> Result<()>;
+
+    /// Set the Substrait plan to execute.
+    ///
+    /// The query can then be executed with [Statement::execute]. For queries
+    /// expected to be executed repeatedly, call [Statement::prepare] first.
+    fn set_substrait_plan(&self, plan: &[u8]) -> Result<()>;
```

Comment:
I'm not familiar with ADBC C api, but aren't these two required? And calling any other function before calling either of these two results in an error?

I would expect these two to be `fn new_sql(...) -> Self` and `fn new_substrait(...) -> Self`, but somehow associated with `Connection`. Maybe something like this:

```rust

enum Query<'a> {
    SQL(&'a str),
    Substrait(&'a [u8]),
}

trait Connection {
    ...

    fn new_statement(&self, query: Query) -> Result<Self::StatementType>;
}
```

Reply:
Overall I agree with you that we need to find a good balance between being idiomatic and close to the C API. That means that trade-offs will need to be made.

1. According to my understanding of the C API, you can use a statement without ever calling `StatementSetSqlQuery` or `StatementSetSubstraitPlan`. For instance, I you want to ingest some data you will set some options on the statement, bind some data with `StatementBind` and finally call `StatementExecuteQuery` (even though there is no query per se).

2. If we go this route, we cannot reuse a statement with a different query (it might not be a big deal since we can always create a fresh statement).

I think we can live with 2, but 1 seems a bit harder. We could pass an `Option<Query>` for the specific case of data ingestion... Maybe there exist some other cases that won't fit your proposed change, I need to think more about it.

Reply:
You are right, C API supports not setting sql query or substrait plan. Although it does that by setting an option on the connection:

https://github.com/apache/arrow-adbc/blob/main/adbc.h#L735-L755

I don't think this is a good design, I would prefer something more explicit, similar to the Java API: https://arrow.apache.org/adbc/current/java/api/org/apache/arrow/adbc/core/AdbcConnection.html#bulkIngest(java.lang.String,org.apache.arrow.adbc.core.BulkIngestMode)

```rust
enum Operation<'a> {
    QuerySQL(&'a str),
    QuerySubstrait(&'a [u8]),
    BulkIngest { target_table_name: &'a str },
}

trait Connection {
    ...

    fn new_statement(&self, op: Operation) -> Result<Self::StatementType>;
}
```

I can also forgo 2, creating a new statement should not be expensive.

Reply:
> See here for instance: https://arrow.apache.org/adbc/current/format/specification.html#bulk-data-ingestion

Ahhh, this is the part of the spec that I was missing!

I like your proposal. It is a good trade off between the two extremes.

It would also not be hard to make a wrapper API that does not use options for such stuff at all and has  idiomatic method structure that cannot be misused.

Reply:
Yes I was thinking of that too: we could build a higher level API which is less error-prone, more idiomatic and that does not have the constraint to accomodate the C API (it will probably not implement all the possible semantics of the C API though).

In that case, this API must only be used by consumer of Rust driver (driver implementors must implement the low-level API to be exportable to C).

Let's keep that in mind for the future.

Reply:
> In that case, this API must only be used by consumer of Rust driver (driver implementors must implement the low-level API to be exportable to C).

I don't understand what you mean here.

Are you saying that implementations of ADBC drivers in Rust will implement this (low-level) API, so it can be accessible via ffi by C code?

Reply:
This is hypothetical but here is my intuition at the moment:
- The low-level API can accomodate all the semantics of the C API
- If your driver implement this low-level API, you can automatically (thanks to the driver exporter) build it as an object file (static or dynamic) that other languages can use (via their own driver manager).
- The high-level API is built on the low-level API and is more idiomatic and less error-prone BUT it cannot be shared with other languages (it might possible to automatically implement this high-level API from the low-level API, ala driver exporter) and probably cannot perform all operations made possible by the C API.

From a driver implementor perspective you should first implement the low-level API, so that you can share your driver with others, which is kind of the point of ADBC. If it's possible to automatically implement the high-level API then this is a win-win. If you decide to only implement the high-level API, only Rust users will be able to use it. 

From a driver consumer perspective you should almost always use the high-level API since it's more ergonomic (except if you need some semantics that's not availabe in the high-level API).

Reply:
I've added the `new_ingest_statement` method, please take a look!

Reply:
Ok, great, we are on the same page.

Now that we also have this idea for a high-level API, there is less need for `new_ingest_statement`, since this API is not meant to be consumer-facing.